python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java

```diff
@@ -45,37 +49,77 @@
 
   private final FileLinesContextFactory fileLinesContextFactory;
 
+  private boolean seenFirstToken;
+
+  private final boolean enableNoSonar;
+
+  private final boolean ignoreHeaderComments;
+
+  private Set<Integer> noSonar = Sets.newHashSet();
   private Set<Integer> linesOfCode = Sets.newHashSet();
   private Set<Integer> linesOfComments = Sets.newHashSet();
   private final FileSystem fileSystem;
   private final Map<InputFile, Set<Integer>> allLinesOfCode;
 
-  public FileLinesVisitor(FileLinesContextFactory fileLinesContextFactory, FileSystem fileSystem, Map<InputFile, Set<Integer>> linesOfCode) {
+  public FileLinesVisitor(
+      FileLinesContextFactory fileLinesContextFactory,
+      FileSystem fileSystem,
+      Map<InputFile, Set<Integer>> linesOfCode,
+      boolean enableNoSonar,
+      boolean ignoreHeaderComments) {
     this.fileLinesContextFactory = fileLinesContextFactory;
     this.fileSystem = fileSystem;
     this.allLinesOfCode = linesOfCode;
+    this.enableNoSonar = enableNoSonar;
+    this.ignoreHeaderComments = ignoreHeaderComments;
+  }
+
+  @Override
+  public void visitFile(AstNode astNode) {
+    noSonar = new HashSet<>();
+    linesOfComments = new HashSet<>();
+    seenFirstToken = false;
   }
 
+  /**
+   * Gets the lines of codes and lines of comments (with character #).
+   * Does not get the lines of docstrings. 
+   */
   @Override
   public void visitToken(Token token) {
     if (token.getType().equals(GenericTokenType.EOF)) {
       return;
     }
 
     if (!token.getType().equals(PythonTokenType.DEDENT) && !token.getType().equals(PythonTokenType.INDENT) && !token.getType().equals(PythonTokenType.NEWLINE)) {
-      /* Handle all the lines of the token */
+      //  Handle all the lines of the token
       String[] tokenLines = token.getValue().split("\n", -1);
       for (int line = token.getLine(); line < token.getLine() + tokenLines.length; line++) {
         linesOfCode.add(line);
       }
     }
 
-    List<Trivia> trivias = token.getTrivia();
-    for (Trivia trivia : trivias) {
-      if (trivia.isComment()) {
-        linesOfComments.add(trivia.getToken().getLine());
+    if (!ignoreHeaderComments || seenFirstToken) {
+      for (Trivia trivia : token.getTrivia()) {
+        if (trivia.isComment()) {
+          String[] commentLines = getContext().getCommentAnalyser().getContents(trivia.getToken().getOriginalValue())
+            .split("(\r)?\n|\r", -1);
+          int line = trivia.getToken().getLine();
+
+          for (String commentLine : commentLines) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)